_drafts/git-bisect.md

```diff
@@ -0,0 +1,585 @@
+the following:
+
+* If the midpoint commit was good, you know the commit that introduced the
+  change is between your midpoint and the initial “bad” commit. 
+* If the midpoint commit was bad, you know the commit that introduced the change
+is between your midpoint and the initial “good” commit.
+
+As a result, you have now effectively cut the whole search space in half, and
+can continue to the midpoint of the remaining search space. This procedure
+repeats on to commit E, to commit G, and finally to either F or H depending on
+whether G is good or bad. Once you’ve finished searching, you have isolated the
+commit that introduced the undesired behavior, and can finally see what
+happened.
+
+### Typical/Important Commands
+There’s quite a few commands within the git bisect world, but here’s the bare
+minimum you would need for a bisect.
+
+* `git bisect start`: puts git into bisect mode, awaiting further bisect
+commands
+* `git bisect bad <commit hash or tag>`: Informs bisect which commit is known
+to be bad
+* `git bisect good <commit hash or tag>`: Informs bisect which commit is known
+to be good
+* `git bisect reset`: exits bisect mode, returns you to normal git mode.
+
+## Let's Go Hunting! A Real Example of Bisecting
+Time to actually bisect through my code and figure out why my flashing script
+isn’t working. Let’s go through this bisect and see what happened.
+
+### Setting Up the Bisect
+Let’s look at my commit history first. I know this was good at v1.0, and bad at
+HEAD (AKA, my latest commit), so let’s get a short list of everything in
+between.
+
+```
+>>> git log --pretty=oneline v1.0~1..HEAD
+e1952a1c Comment flasher files
+51ed84b6 Comment main.c
+b3a10580 Helpful python error message and traceback
+f21a4c55 (tag v1.2) Add comments and static-ize flasher.c
+47a89108 Actually improved performance
+0afd538f Reverting performance improvements that made things slower
+84c98ac9 Performance improvements
+26338040 Refactor python script with classes
+54526f51 (tag v1.1) Note the work that still needs to be done in the readme
+9f0d9a91 Move flash driver to non-memory-mapped
+e717c312 Repo linting changes
+a2b8a742 More misc. changes
+e0c258c0 Misc. Changes
+cecc054a Integrate all the things
+acf78303 (tag v1.0) Add initialization procedure
+```
+
+Cool, not too much here, just 15 commits. Let’s set up the bisect using the
+commands we just learned
+
+```
+git bisect start
+git bisect bad HEAD (same as git bisect e1952a1c)
+git bisect good v1.0 (same as git bisect acf78303)
+```
+
+### Running the Bisect
+Now we see this output:
+
+```
+Bisecting: 6 revisions left to test after this (roughly 3 steps)
+[26338040b292e110c2086de6bc42d1c232005a58] Refactor python script with classes
+```
+
+In a very helpful output, git is telling us how many different commits we still
+need to run through, and is counting down about how many test steps we have
+left. The view of our bisect looks something like this right now:
+
+```
+RED
```

Comment:
Couldn't figure out how to do color in Markdown, but any suggestions are welcome. For now, I just labeled the color before what I want the next set of commits to be.